libcontainer/nsenter/cloned_binary.c

```diff
@@ -192,49 +190,149 @@ static int fetchve(char ***argv)
+
+	/*
+	 * Try O_TMPFILE to avoid races where someone might snatch our file. Note
+	 * that O_EXCL isn't actually a security measure here (since you can just
+	 * fd re-open it and clear O_EXCL).
+	 */
+	*fdtype = EFD_TMPFILE;
+	if (*fdtype & allowed) {
+		fd = open("/tmp", O_TMPFILE | O_EXCL | O_RDWR | O_CLOEXEC, 0700);
+		if (fd >= 0) {
+			struct stat statbuf = {};
+			bool working_otmpfile = false;
+
+			/*
+			 * open(2) ignores unknown O_* flags -- yeah, I was surprised when I
+			 * found this out too. As a result we can't check for EINVAL. However,
+			 * if we get nlink != 0 (or EISDIR) then we know that this kernel
+			 * doesn't support O_TMPFILE.
+			 */
+			if (fstat(fd, &statbuf) >= 0)
+				working_otmpfile = (statbuf.st_nlink == 0);
+
+			if (working_otmpfile)
+				return fd;
+
+			/* Pretend that we got EISDIR since O_TMPFILE failed. */
+			close(fd);
+			errno = EISDIR;
+		}
+		if (errno != EISDIR)
+			goto err;
+	}
+
+	/*
+	 * Our final option is to create a temporary file the old-school way, and
+	 * then unlink it so that nothing else sees it by accident.
+	 */
+	*fdtype = EFD_FILE;
+	if (*fdtype & allowed) {
+		fd = mkostemp(template, O_CLOEXEC);
+		if (fd >= 0) {
+			if (!unlink(template))
+				return fd;
+			close(fd);
+		}
+	}
+
+err:
+	*fdtype = EFD_NONE;
+	return -1;
+}
+
+static int seal_execfd(int *fd, int fdtype)
+{
+	switch (fdtype) {
+	case EFD_MEMFD:
+		return fcntl(*fd, F_ADD_SEALS, RUNC_MEMFD_SEALS);
+	case EFD_TMPFILE: /* fall through */
+	case EFD_FILE: {
+		/* Need to re-open "memfd" as an O_PATH to avoid execve(2) giving -EXTBSY. */
+		int newfd;
+		char *fdpath = NULL;
+
+		if (fchmod(*fd, 0100) < 0)
+			return -1;
+
+		while (asprintf(&fdpath, "/proc/self/fd/%d", *fd) < 0)
+			/* noop */;
+		newfd = open(fdpath, O_PATH | O_CLOEXEC);
+		free(fdpath);
+		if (newfd < 0)
+			return -1;
+
+		close(*fd);
+		*fd = newfd;
+		return 0;
+	}
+	default:
+	   break;
+	}
+	return -1;
+}
+
 static int clone_binary(void)
 {
 	int binfd, memfd;
-	ssize_t sent = 0;
+	struct stat statbuf = {};
+	size_t sent = 0;
+	int fdtype = EFD_NONE, allowed = EFD_ALL;
 
-#ifdef HAVE_MEMFD_CREATE
-	memfd = memfd_create(RUNC_MEMFD_COMMENT, MFD_CLOEXEC | MFD_ALLOW_SEALING);
-#else
-	memfd = open("/tmp", O_TMPFILE | O_EXCL | O_RDWR | O_CLOEXEC, 0711);
-#endif
+	if (secure_getenv("_LIBCONTAINER_DISABLE_MEMFD_CLONE"))
+		allowed &= ~EFD_MEMFD;
+
+	memfd = make_execfd(&fdtype, allowed);
 	if (memfd < 0)
 		return -ENOTRECOVERABLE;
 
 	binfd = open("/proc/self/exe", O_RDONLY | O_CLOEXEC);
 	if (binfd < 0)
 		goto error;
 
-	sent = sendfile(memfd, binfd, NULL, RUNC_SENDFILE_MAX);
-	close(binfd);
-	if (sent < 0)
-		goto error;
+	if (fstat(binfd, &statbuf) < 0)
+		goto error_binfd;
 
-#ifdef HAVE_MEMFD_CREATE
-	int err = fcntl(memfd, F_ADD_SEALS, RUNC_MEMFD_SEALS);
-	if (err < 0)
+	while (sent < statbuf.st_size) {
+		int n = sendfile(memfd, binfd, NULL, statbuf.st_size - sent);
+		if (n < 0)
+			goto error_binfd;
+		sent += n;
+	}
+	close(binfd);
```

Comment:
`sendfile()` doesn't necessarily work correctly on all filesystems. If you're not on a tmpfs aka using `memfd_create()` you should probably have a `fd_to_fd()` or something. But *shrug* :)

Reply:
(Referring back to our chat before. :))

Reply:
I will test it with some other filesystems, but as I discussed in our chat unless your filesystem defines a `f_op->splice_write` then it should work just as well as it would anywhere (and it's all handled in the VFS with `current->splice_pipe`.

Reply:
I'm already tracking this thing down.

Reply:
So sendfile() works but I've added a fallback `fd_to_fd()`  in our code.

Reply:
For the record, the issue was that AppArmor blocks writing to deleted files. If someone complains about this, I can fix it later but I don't feel the need to include `fd_to_fd()` at the moment.

Reply:
I've added `fd_to_fd()`.